data_structures/rb_tree2.cpp

```diff
@@ -0,0 +1,1686 @@
+ *
+ */
+struct RBTNode {
+    /// nodes are red by default
+    /// null/root/red's child nodes are recolored to black
+    uint16_t color = COLOR_RED;
+    /// has int value to represent node itself
+    int val;
+
+    /// has a parent and left/right child nodes
+    RBTNode* parent = nullptr;
+    RBTNode* left = nullptr;
+    RBTNode* right = nullptr;
+
+    bool treatAsNull = false;
+};
+
+class RedBlackTree {
+ public:
+        /**
+         * @brief Constructor - Initializes new Red Black Tree
+         *
+         */
+        RedBlackTree() {
+            root = nullptr;
+            numItems = 0;
+        }
+
+        /**
+         * @brief Deconstructor - calls helper PostDelete to
+         * free Red Black Tree memory
+         *
+         */
+        ~RedBlackTree() { PostDelete(root); }
+
+        // Helper for deconstructor - delete nodes in post-order
+        void PostDelete(RBTNode* node) {
+            // delete left then right (recursively)
+            if (node != nullptr) {
+                PostDelete(node->left);
+                PostDelete(node->right);
+            }
+            // delete parent
+            delete node;
+        }
+
+        /**
+         * @brief Constructs a new copy of the Red Black Tree passed
+         * to this function.
+         *
+         * @param rbt the Red Black Tree to be copied
+         */
+        RedBlackTree(const RedBlackTree& rbt) {
+            // initialize tree first
+            this->root = nullptr;
+            this->numItems = 0;
+
+            // insert nodes from passed tree in the same order
+            // they were inserted into the original tree
+            for (int n : rbt.nodes_order_added) {
+                this->Insert(n);
+            }
+        }
+
+        /**
+         * @brief returns size of red black tree
+         *
+         */
+        uint64_t Size() {
+            return numItems;
+        }
+
+        /**
+         * @brief Tree in infix notation
+         *
+         * @return string
+         */
+        string ToInfixString() const { return ToInfixString(root); }
+
+        /**
+         * @brief Tree in prefix notation
+         *
+         * @return string
+         */
+        string ToPrefixString() const { return ToPrefixString(root); }
+
+        /**
+         * @brief Tree in postfix notation
+         *
+         * @return string
+         */
+        string ToPostfixString() const { return ToPostfixString(root); }
+
+        /**
+         * @brief Inserts new node into tree with the
+         * value passed.
+         *
+         * @param value value of the new node
+         */
+        void Insert(int value) {
+            // check for duplicate keys
+            if (this->Contains(value)) {
+                throw invalid_argument(
+                    "Insert(): Node is already in tree"
+                    ", duplicate nodes not allowed.");
+            }
+            RBTNode *node_to_insert = new RBTNode();
```

Comment:
Might wanna use `auto` here

Reply:
Why here? Just curious, and thanks again for the notes on my code here!

Reply:
I usually have two reasons for using auto instead of using the actual type:
- when the type is already known from the right hand side of the operand (in this case `new RBTNode()` is a good indicator)
- when the type is kind of irrelevant (there are certain cases of its abuse but I won't go into the details of it here).